consensus/scp/tests/mock_network/mod.rs

```diff
@@ -230,151 +264,131 @@ enum SCPNodeTaskMessage {
-                                    incoming_msgs.push(msg);
+                                    incoming_msg_option = Some(msg);
                                 }
 
-                                // Request to stop thread
+                                // Stop the thread
                                 SCPNodeTaskMessage::StopTrigger => {
                                     break 'main_loop;
                                 }
                             },
                             Err(_) => {
                                 // Yield to other threads when we don't get a new message
-                                // This improves performance significantly.
                                 std::thread::yield_now();
                             }
                         };
 
-                        let incoming_msgs_count = incoming_msgs.len();
-
-                        if !(incoming_msgs_count == 0 || incoming_msgs_count == 1) {
-                            log::error!(logger, "incoming_msgs_count > 1");
-                            // panic
-                            assert!(incoming_msgs_count == 0 || incoming_msgs_count == 1);
-                        }
-
-                        // Process values submitted to our node
-                        if (nominated_values < test_options.max_values_per_slot)
+                        // Nominate pending values submitted to our node
+                        if (slot_nominated_values.len() < max_pending_values_to_nominate)
                             && !pending_values.is_empty()
                         {
-                            let mut vals = pending_values.iter().cloned().collect::<Vec<String>>();
-                            vals.sort();
-                            vals.truncate(test_options.max_values_per_slot - nominated_values);
-                            nominated_values += vals.len();
-
-                            let outgoing_msg: Option<Msg<String>> = {
-                                thread_local_node
-                                    .lock()
-                                    .expect("lock failed on node nominating value")
-                                    .propose_values(BTreeSet::from_iter(vals))
-                                    .expect("node.nominate() failed")
-                            };
+                            let mut values: Vec<String> = pending_values.iter().cloned().collect();
+                            values.sort();
+                            values.truncate(max_pending_values_to_nominate);
 
-                            if let Some(outgoing_msg) = outgoing_msg {
-                                (broadcast_msg_fn)(logger.clone(), outgoing_msg);
-                                total_broadcasts += 1;
+                            // mc_common::HashSet does not support extend because of our enclave-safe HasherBuilder
+                            let mut values_to_nominate: HashSet<String> =
```

Comment:
If you use `BTreeSet` here then you won't have to then convert it into a BTreeSet a few lines below.

Reply:
I updated this to use a BTreeSet as you suggested, and reconfirmed the performance improvement that I get by reducing the rate of proposing values to slot:


```
type	with	without optimization
c1	1.101	1.094
c2	1.104	8.941
c3	2.049	10.874
c4	2.605	17.448
c5	2.515	24.399
c6	3.666	39.877
m1k0	1.098	1.103
m2k1	1.099	2.903
m3k1	1.137	5.493
m3k2	1.099	12.607
m4k3	1.518	18.279
m5k3	2.758	23.477
m5k4	2.762	23.525
```

I really think we should see if a similar optimization can help in byzantine_ledger... @eranrund

Reply:
Looking into this further, I think the crate tests switched `pending_values` to a hashset at some point in the past and that this actually created the problems this optimization fixes. I switched back to keeping pending values in a vector so that the continued proposals do not inadvertently introduce new values...  

The thing we want to avoid is slowly feeding new values to a slot once consensus is underway -- it's better to queue up new values and propose them all at once to a future slot rather than call `propose` multiple times per slot. 

Ideally the scp integration tests would use the same worker threads as byzantine ledger.

Reply:
In ByzanitneLedger's worker thread, I think the intention is that `pending_values` is a FIFO queue of values (i.e. transactions) submitted directly by clients to that node. I think the goal there is to maintain fairness, where each node is trying to nominate the transactions it has received from clients in the order that it received them.

Requiring each node to sort pending values first (and then try to nominate them in that order) might improve throughput, but maybe at the expense of max latency or fairness. For example, if transactions are sorted by TxHash ascending, a transaction that happens to have a very high TxHash might get repeatedly passed up for nomination. Or, a clever client who wants "expedited" transactions could generate a number of equivalent transactions with different hashes (e.g. same inputs, same outputs, different nonces) then "jump to the head of the line" by submitting the one with the lowest TxHash.